ethcore/native_contracts/generator/src/lib.rs

```diff
@@ -0,0 +1,346 @@
+// as a tuple, and the second gives code to get that tuple from a deque of tokens.
+//
+// produce output type of the form (type_1, type_2, ...) without trailing comma.
+// produce code for getting this output type from `outputs: VecDeque<Token>`, where
+// an `Err(String)` can be returned.
+//
+// returns any unsupported param type encountered.
+fn output_params_codegen(outputs: &[ParamType]) -> Result<(String, String), ParamType> {
+	let mut output_type = "(".to_string();
+	let mut decode_outputs = "Ok((".to_string();
+
+	for (index, output) in outputs.iter().cloned().enumerate() {
+		let rust_type = rust_type(output.clone())?;
+
+		output_type.push_str(&rust_type);
+
+		decode_outputs.push_str(&format!(
+			r#"
+				outputs
+					.pop_front()
+					.and_then(|output| {{ {} }})
+					.ok_or_else(|| "Wrong output type".to_string())?
+			"#,
+			detokenize("output", output)
+		));
+
+		// don't append trailing commas for the last element
+		// so we can reuse the same code for single-output contracts,
+		// since T == (T) != (T,)
+		if index < outputs.len() - 1 {
+			output_type.push_str(", ");
+			decode_outputs.push_str(", ");
+		}
+	}
+
+	output_type.push_str(")");
+	decode_outputs.push_str("))");
+	Ok((output_type, decode_outputs))
+}
+
+// create code for an argument type from param type.
+fn rust_type(input: ParamType) -> Result<String, ParamType> {
+	Ok(match input {
+		ParamType::Address => "util::Address".into(),
+		ParamType::FixedBytes(len) if len <= 32 => format!("util::H{}", len * 8),
+    	ParamType::Bytes | ParamType::FixedBytes(_) => "Vec<u8>".into(),
+    	ParamType::Int(width) => match width {
+			8 | 16 | 32 | 64 => format!("i{}", width),
+			_ => return Err(ParamType::Int(width)),
+		},
+    	ParamType::Uint(width) => match width {
+			8 | 16 | 32 | 64 => format!("u{}", width),
+			128 | 160 | 256 => format!("util::U{}", width),
+			_ => return Err(ParamType::Uint(width)),
+		},
+    	ParamType::Bool => "bool".into(),
+    	ParamType::String => "String".into(),
+		ParamType::Array(kind) => format!("Vec<{}>", rust_type(*kind)?),
+		other => return Err(other),
+	})
+}
+
+// create code for tokenizing this parameter.
+// returns (needs_mut, code), where needs_mut indicates mutability required.
+// panics on unsupported types.
+fn tokenize(name: &str, input: ParamType) -> (bool, String) {
+	let mut needs_mut = false;
+	let code = match input {
+		ParamType::Address => format!("Token::Address({}.0)", name),
+		ParamType::Bytes => format!("Token::Bytes({})", name),
+		ParamType::FixedBytes(len) if len <= 32 =>
+			format!("Token::FixedBytes({}.0.to_vec())", name),
+		ParamType::FixedBytes(len) => {
+			needs_mut = true;
+			format!("{}.resize({}, 0); Token::FixedBytes({})", name, len, name)
+		}
+		ParamType::Int(width) => match width {
+			8 => format!("let mut r = [0xff; 32]; r[31] = {}; Token::Int(r)", name),
+			16 | 32 | 64 =>
+				format!("let mut r = [0xff; 32]; BigEndian::write_i{}(&mut r[{}..], {}); Token::Int(r))",
```

Comment:
@debris I'm not all that familiar with ethabi -- but assuming the int width is fixed, does it matter what the left side of the array holds? is `0xff` right?